### DISCUSSION TITLE
Restart browser for each feature instead of for each Scenario

### DISCUSSION THREAD

**Discussion by ahmed-al-sadoon** [serenity-js/serenity-js#2277] [How to... ?] [answered]
Hi there,
I'm using SerenityJS with Cucumber and Playwright and I have a feature file with multi scenarios, I would like that the browser restartet for each feature instead of each scenario like in Serenity-BDD "**_restart.browser.for.each = feature_**". 
Is it posssible in SerenityJS?

**Comment by jan-molak** [comment, marked as answer]
Hi @ahmed-al-sadoon! 

[Cucumber.js hooks](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/hooks.md) work in the scope of an entire test suite (`BeforeAll`/`AfterAll`), scenario (`Before`/`After`), or a step (`BeforeStep`/`AfterStep`). Cucumber doesn't support hooks that would be executed before/after a feature because of the reasons described in https://github.com/cucumber/cucumber-js/issues/914#issuecomment-325559474

Also, it's usually better to keep your test scenarios isolated, so that the state from one scenario doesn't leak to another.

Having said that, if you've considered the above and concluded that you still want to do it, here's how.

## Integrating with Playwright

Serenity/JS actors integrate with Playwright using the ability to [`BrowseTheWebWithPlaywright`](https://serenity-js.org/api/playwright/class/BrowseTheWebWithPlaywright/).

There are two ways to inject Playwright API into this ability that work slightly differently:
- [`BrowseTheWebWithPlaywright.using(browser, options)`](https://serenity-js.org/api/playwright/class/BrowseTheWebWithPlaywright/#using) creates a new [Playwright browser context](https://playwright.dev/docs/api/class-browsercontext) and [page](https://playwright.dev/docs/api/class-page) the first time the ability is used. When the actors are [dismissed](https://serenity-js.org/api/core/class/Actor/#dismiss) - it closes any open pages, as well as the context itself. Note that actors instantiated in a `Before` hook or a step definition are dismissed when the scenario finishes, while actors instantiated in `BeforeAll` are dismissed when all the scenarios finish. This is the setup recommended for most Cucumber integrations and [offered in the default template](https://github.com/serenity-js/serenity-js-cucumber-playwright-template/blob/9777ac18632ea980d421fe46a65f4e6e4f94e3f2/features/support/serenity.config.ts#L21-L63)
- [`BrowseTheWebWithPlaywright.usingPage(page, options)`](https://serenity-js.org/api/playwright/class/BrowseTheWebWithPlaywright/#usingPage) - assumes that the test runner, such as Playwright Test, is responsible for managing the browser context and creating the initial `page`, so will not create a new context or new page, but reuse the received object. In this case, when the actor is dismissed, any additional pages/new tabs opened while the actor was active will be closed, but the initial page and context are left intact. This setup is recommended and default for Playwright Test runner.

Note that both the above setups keep the actual `browser` process running for the duration of your entire test run and operate on [browser context](https://playwright.dev/docs/api/class-browsercontext), which is much cheaper and faster to create and destroy than restarting the whole browser.

## Reusing the same `page` between scenarios

> [!CAUTION]
> Reusing the same `page` object between test scenarios will ensure that Serenity/JS will not attempt to close it, but will likely lead to state leakage and flaky tests if your test scenarios rely on other scenarios to bring the system under test to a desired state.
> A common anti-pattern is to have a feature file where the first scenario logs the test user in, and the second scenario assumes that the user is already logged in, or similar. Please try to avoid that and keep your scenarios independent.

To reuse the same `page` between test scenarios, you can follow the pattern proposed in the [Playwright Test documentation](https://playwright.dev/docs/test-retries#reuse-single-page-between-tests).

So instead of injecting the browser, as [recommended in the Serenity/JS Cucumber Playwright template](https://github.com/serenity-js/serenity-js-cucumber-playwright-template/blob/9777ac18632ea980d421fe46a65f4e6e4f94e3f2/features/support/serenity.config.ts#L21C1-L63C3):

```ts
import { AfterAll, BeforeAll, setDefaultTimeout } from '@cucumber/cucumber';
import { configure } from '@serenity-js/core';
import * as playwright from 'playwright';

let browser: playwright.Browser;

BeforeAll(async () => {

    browser = await playwright.chromium.launch({
        headless: true,
    });

    configure({
        actors: Cast.where(actor => {
          return actor.whoCan(BrowseTheWebWithPlaywright.using(browser)),
        }),
    });
});

AfterAll(async () => {
    if (browser) {
        await browser.close();
    }
});
```

You create a context yourself, instantiate the initial `page`, and inject that instead:

```ts
import { AfterAll, BeforeAll, setDefaultTimeout } from '@cucumber/cucumber';
import { configure } from '@serenity-js/core';
import * as playwright from 'playwright';

let browser: playwright.Browser;
let page: playwright.Page;

BeforeAll(async () => {

    browser = await playwright.chromium.launch({
        headless: true,
    });
    
    page = await browser.newPage();

    configure({
        actors: Cast.where(actor => {
          return actor.whoCan(BrowseTheWebWithPlaywright.usingPage(page)),
        }),
    });
});

AfterAll(async () => {
    if (page) {
        await page.close();
    }

    if (browser) {
        await browser.close();
    }
});
```

This way, all your test scenarios will reuse the same page instance, and Serenity/JS will not attempt to close it.

Since you're now in full control of the page, you can also introduce `Before` and `After` hooks that manipulate the `page` and recreate it when needed.

## How to work around the lack of before/after feature hooks

The Cucumber.js team [actively discourages the idea of before/after feature hooks](https://github.com/cucumber/cucumber-js/blob/main/docs/faq.md#why-arent-there-beforefeature-and-afterfeature-hooks), suggesting for you to use [`@tags`](https://cucumber.io/docs/cucumber/api/?lang=javascript#tags) instead. I agree with this view and recommend that test scenarios should be independent.

**Reply by ahmed-al-sadoon** [reply]
Hi @jan-molak 
Thanks alot for your reply and your advice
